lightningd/plugin.c

```diff
@@ -2080,43 +2086,67 @@ static void plugin_shutdown_timeout(struct lightningd *ld)
 	io_break(ld->plugins);
 }
 
-void shutdown_plugins(struct lightningd *ld)
+void shutdown_plugins(struct lightningd *ld, bool first)
 {
 	struct plugin *p, *next;
 
-	/* This makes sure we don't complain about important plugins
-	 * vanishing! */
-	ld->plugins->shutdown = true;
-
-	/* Tell them all to shutdown; if they care. */
+	/* Mark the set of plugins we want to shutdown in this call */
 	list_for_each_safe(&ld->plugins->plugins, p, next, list) {
-		/* Kill immediately, deletes self from list. */
-		if (!notify_plugin_shutdown(ld, p))
+		if (first && plugin_registered_db_write_hook(p))
+			continue;
+
+		/* don't complain about important plugins vanishing and stop sending
+		 * it any (other) notifications */
+		plugin_set_state(p, SHUTDOWN);
+
+		/* notify subscribed ones, others are killed immediately */
+		if (notify_plugin_shutdown(ld, p))
+			continue;
+		else
 			tal_free(p);
 	}
 
-	/* If anyone was interested in shutdown, give them time. */
-	if (!list_empty(&ld->plugins->plugins)) {
+	/* Remaining plugins in the set have 30s or 5s to self-terminate */
+	if (plugins_any_in_state(ld->plugins, SHUTDOWN)) {
 		struct oneshot *t;
-
-		/* 30 seconds should do it. */
+		void *ret;
 		t = new_reltimer(ld->timers, ld,
-				 time_from_sec(30),
+				 time_from_sec(first ? 30 : 5),
 				 plugin_shutdown_timeout, ld);
 
+		/* Destructor of plugins could've called io_break, which makes the next
+		 * io_loop return immediately, so we explicitly reset it */
+		io_break(ld);
+		ret = io_loop(NULL, NULL);
+		assert(ret == ld);
+
 		io_loop_with_timers(ld);
 		tal_free(t);
 
-		/* Report and free remaining plugins. */
-		while (!list_empty(&ld->plugins->plugins)) {
-			p = list_pop(&ld->plugins->plugins, struct plugin, list);
+		/* Report and free plugins that didn't self-terminate */
+		list_for_each_safe(&ld->plugins->plugins, p, next, list) {
+			if (p->plugin_state != SHUTDOWN)
+				continue;
+
 			log_debug(ld->log,
-				  "%s: failed to shutdown, killing.",
+				  "%s: failed to self-terminate, killing.",
 				  p->shortname);
 			tal_free(p);
 		}
 	}
 
-	/* NULL stops notifications trying to access plugins. */
-	ld->plugins = tal_free(ld->plugins);
+	/* In 2nd call all should be gone, close the db */
+	if (!first) {
+		assert(list_empty(&ld->plugins->plugins));
+		ld->plugins = tal_free(ld->plugins);
+		tal_free(ld->wallet);
```

Comment:
Oh, now I catch your idea. it makes sense, also because the plugin can be a backup plugin, so we should stop writing on the database.

Concept ack, and Great Work!